BraveShared/Favicons/FaviconFetcher.swift

```diff
@@ -0,0 +1,146 @@
+          guard let self = self, !self.isCancelled else {
+            completion?(nil)
+            return
+          }
+          
+          // A favicon was fetched
+          if let favicon = favicon {
+            storeInCache(favicon, for: url, persistent: persistent)
+          }
+          
+          completion?(favicon)
+        }
+      }
+    }
+    
+    cancellable.execute(on: .main)
+    return cancellable
+  }
+  
+  public class Cancellable {
+    private var workItem: DispatchWorkItem?
+    private var task: (Cancellable, BundledFaviconImageRenderer, FaviconRenderer) -> Void
+    private var bundledRenderer: BundledFaviconImageRenderer?
+    private var faviconRenderer: FaviconRenderer?
+    
+    public var isCancelled: Bool {
+      workItem?.isCancelled ?? true
+    }
+    
+    public func cancel() {
+      workItem?.cancel()
+      workItem = nil
+    }
+    
+    fileprivate init(task: @escaping (Cancellable?, BundledFaviconImageRenderer?, FaviconRenderer?) -> Void) {
+      self.task = task
+      self.bundledRenderer = BundledFaviconImageRenderer()
+      self.faviconRenderer = FaviconRenderer()
+      
+      self.workItem = DispatchWorkItem { [weak self] in
+        task(self, self?.bundledRenderer, self?.faviconRenderer)
+      }
+    }
+    
+    fileprivate func execute(on queue: DispatchQueue) {
+      if let workItem = workItem {
+        queue.async(execute: workItem)
+      }
+    }
+  }
+  
+  private static func cacheURL(for url: URL) -> URL {
+    // Some websites still only have a favicon for the FULL url including the fragmented parts
+    // But they won't have a favicon for their domain
+    // In this case, we want to store the favicon for the entire domain regardless of query parameters or fragmented parts
+    // Example: `https://app.uniswap.org/` has no favicon, but `https://app.uniswap.org/#/swap?chain=mainnet` does.
+    return URLOrigin(url: url).url ?? url
+  }
+  
+  private static func storeInCache(_ favicon: Favicon, for url: URL, persistent: Bool) {
+    // Do not cache non-persistent icons
+    // Do not cache monogram icons
+    if !persistent && !favicon.isMonogramImage {
+      guard let data = try? JSONEncoder().encode(favicon) else {
+        return
+      }
+      
+      let cachedURL = cacheURL(for: url)
+      SDImageCache.shared.memoryCache.setObject(data, forKey: cachedURL.absoluteString, cost: UInt(data.count))
+      SDImageCache.shared.diskCache.setData(data, forKey: cachedURL.absoluteString)
+    }
+  }
+  
+  private static func getFromCache(for url: URL) -> Favicon? {
+    let cachedURL = cacheURL(for: url)
+    let data = SDImageCache.shared.memoryCache.object(forKey: cachedURL.absoluteString) as? Data ??
+               SDImageCache.shared.diskCache.data(forKey: cachedURL.absoluteString)
+    
+    if let data = data,
+        let favicon = try? JSONDecoder().decode(Favicon.self, from: data) {
+      return favicon
+    }
+    
+    return nil
```

Comment:
1. Same comment as above, don't do optional try, catch error and log
2. Should rearrange to return nil early